lib/upload.ts

```diff
@@ -0,0 +1,215 @@
+
+const schemaUser = z.object({
+  admissionsCycle: z.coerce.number().int().positive(),
+  login: z.string(),
+  role: z.nativeEnum(Role)
+})
+
+// nested schema for upserting an applicant and creating a new application
+const schemaApplication = z.object({
+  applicant: z.object({
+    cid: z.string().length(8, { message: 'CID must be exactly 8 characters' }),
+    ucasNumber: z.string().length(10, { message: 'UCAS number must be exactly 10 digits' }),
+    gender: z.nativeEnum(Gender),
+    firstName: z.string(),
+    surname: z.string(),
+    preferredName: z.string().nullable(),
+    dateOfBirth: z
+      .string()
+      .transform((value) => formatISO(parseDate(value, 'dd/MM/yyyy', new Date()))),
+    email: z.string().email(),
+    primaryNationality: z.string(),
+    otherNationality: z.string().nullable()
+  }),
+  application: z.object({
+    hasDisability: z.preprocess((value) => String(value).toLowerCase() === 'true', z.boolean()),
+    admissionsCycle: z.coerce.number().int().positive(),
+    feeStatus: z.nativeEnum(FeeStatus).optional().default(FeeStatus.UNKNOWN),
+    wideningParticipation: z.preprocess(
+      (value) => String(value).toLowerCase() === 'true',
+      z.boolean()
+    ),
+    applicationDate: z
+      .string()
+      .transform((value) => formatISO(parseDate(value, 'dd/MM/yyyy HH:mm', new Date())))
+  })
+})
+
+// used to insert TMUA grades
+const schemaTMUAScores = z.object({
+  cid: z.string().length(8, { message: 'CID must be exactly 8 characters' }),
+  tmuaPaper1Score: z.coerce.number().min(1).max(9).optional(),
+  tmuaPaper2Score: z.coerce.number().min(1).max(9).optional(),
+  tmuaOverallScore: z.coerce.number().min(1).max(9).optional()
+})
+
+function parseWithSchema<T extends ZodSchema>(
+  objects: unknown[],
+  validationSchema: T
+): { data: z.infer<T>[]; noErrors: number } {
+  const parsedObjects = objects.map((o) => validationSchema.safeParse(o))
+  parsedObjects.forEach((o, i) => {
+    if (!o.success)
+      console.error(
+        `Parsing error on row ${i + 1}: ${o.error.issues.map((issue) => issue.message).join('; ')}`
+      )
+  })
+  const validObjects = parsedObjects.filter((o) => o.success).map((o) => o.data)
+  return {
+    data: validObjects,
+    noErrors: objects.length - validObjects.length
+  }
+}
+
+async function executePromises(promises: Promise<unknown>[]): Promise<number> {
+  const results = await Promise.allSettled(promises)
+  results.forEach((result) => {
+    if (result.status === 'rejected') {
+      if (result.reason instanceof PrismaClientKnownRequestError) {
+        console.error(
+          `${result.reason.code}: ${result.reason.message
+            .trimStart()
+            .split('\n')
+            .filter((line) => !!line)
+            .join('\n')}`
+        )
+      } else if (result.reason instanceof PrismaClientValidationError) {
+        console.error(result.reason.message)
+      } else {
+        console.error(result.reason?.message)
+      }
```

Comment:
```suggestion
      } else {
        console.error(result.reason?.message ?? "Unknown Prisma error.")
      }
```

Reply:
More! MORE!

```suggestion
      } else console.error(result.reason?.message ?? "Unknown Prisma error.")
```

Reply:
addressed in a778cbb4e71a6949125dffea2e028018148129ce